### PR TITLE
box: handle SIGBUS in flightrec (CE Part)

### PR DESCRIPTION
**Pull request by nshy**
Flight recorder uses mmapped file for performance reasons. Unfortunately if for some reason mapping is not possible we got SIGBUS on accessing mmapped memory.

We already handle this issue on flight recorder start in [1]. But if we got SIGBUS in the meanwhile we currently got a crash. Let's just panic.

[1] https://github.com/tarantool/tarantool-ee/commit/e68b84768f3ae76aa9f03d36dd6f0587b884e1bb

EE Part: https://github.com/tarantool/tarantool-ee/pull/344

### CODE REVIEW

File: src/box/box.cc [hunk 1]
```diff
@@ -2727,20 +2745,214 @@ box_set_prepared_stmt_cache_size(void)
...
+	char *tail = &p[SMALL_STATIC_SIZE];
+	char *e = &p[SMALL_STATIC_SIZE];
+	char *head = p;
+
+	int total = 0;
+	(void)total;
+	int size = 0;
+
+#define snprintf_safe(...) SNPRINT(total, snprintf, p, size, __VA_ARGS__)
+#define jnprintf_safe(str) SNPRINT(total, json_escape, p, size, str)
```
Comment (locker) [inline src/box/box.cc:2770]: Nit: Let's undef these two macros as soon as we don't need them anymore.

File: src/lib/core/crash.h [hunk 2]
```diff
@@ -9,23 +9,124 @@
...
+#ifdef HAS_GREG
+	/**
+	 * Registers contents.
+	 */
+	struct crash_greg greg;
+#endif
+	/**
+	 * Timestamp in seconds (realtime).
+	 */
+	time_t timestamp_rt;
```
Comment (locker) [inline src/lib/core/crash.h:103]: Why did you change the type of this field from `uint64_t` to `time_t`?

File: src/main.cc [hunk 3]
```diff
@@ -721,8 +721,22 @@ main(int argc, char **argv)
...
+	/*
+	 * We need to keep clock data locally to report uptime without binding
+	 * to libev and etc. Because we're reporting information at the moment
+	 * when crash happens and we are to be independent as much as we can.
+	 */
+	struct timespec ts;
+	if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
+		box_start_time = ts.tv_sec;
+	else
+		say_syserror("fail to get box_start_time, ignore");
```
Comment (locker) [inline src/main.cc:737]: Nit: Let's change to `panic`? We don't expect a failure here, and if the clock doesn't work, we'll probably panic somewhere else anyway.

File: src/lib/core/crash.h [hunk 4]
```diff
@@ -9,23 +9,124 @@
 extern "C" {
 #endif /* defined(__cplusplus) */
 
-/**
- * PATH_MAX is too big and 2K is recommended
- * limit for web address.
- */
-#define CRASH_FEEDBACK_HOST_MAX 2048
+#if defined(TARGET_OS_LINUX) && defined(__x86_64__)
```
Comment (locker) [inline src/lib/core/crash.h:12]: You need to include `trivia/config.h`, otherwise `TARGET_OS_LINUX` may not be defined.

File: src/lib/core/crash.h [hunk 5]
```diff
@@ -9,23 +9,124 @@
...
- * Initialize crash subsystem.
+#ifdef HAS_GREG
+/*
+ * Values of x86 registers.
  */
-void
-crash_init(const char *tarantool_bin);
+struct crash_greg {
+	/* r8 register value */
+	uint64_t r8;
```
Comment (locker) [inline src/lib/core/crash.h:22]: Nit: Please include `stdint.h` and other headers needed to compile this one.

File: src/lib/core/crash.h [hunk 6]
```diff
@@ -9,23 +9,124 @@
...
+	 */
+	char backtrace_buf[1024];
+#endif
+};
+
+typedef void
+(*crash_callback_f)(struct crash_info *cinfo);
+
+/* Callback to call on crash. */
+extern crash_callback_f crash_callback;
```
Comment (locker) [inline src/lib/core/crash.h:129]: Nit: Please improve on the comment: can it be NULL? what's it set to by default?

File: src/main.cc [hunk 7]
```diff
@@ -721,8 +721,22 @@ main(int argc, char **argv)
...
+
+	crash_callback = box_crash_callback;
```
Comment (locker) [inline src/main.cc:739]: IMO, better set it in `box_init` - anyway, feedback won't be enabled until the box is configured.

File: src/box/box.cc [hunk 8]
```diff
@@ -2925,6 +2925,13 @@ box_feedback_report_crash(struct crash_info *cinfo)
 void
 box_crash_callback(struct crash_info *cinfo)
 {
+	if (cinfo->signo == SIGBUS &&
+	    flightrec_is_mmapped_address(cinfo->siaddr)) {
+		fprintf(stderr, "error accessing flightrec file\n");
+		fflush(stderr);
+		_exit(EXIT_FAILURE);
```
Comment (locker) [inline src/box/box.cc:2932]: The callback is invoked after dumping the stack trace to stderr. I think it should be called before, because a stack trace looks pretty scary to users. AFAIR we agreed to *replace* the default crash callback instead of adding an extra handler and add make the default handler (the one that dumps the stack trace) available outside crash.c so that it can be reused in box.cc.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/main.cc (hunk 3 above)
Comment (locker) [reply src/main.cc:737]: Agreed to leave as is, because the old code used `pr_syserr`.

Let's use a more human-readable error message (without reference to `box_start_time`).

File: src/box/box.cc [hunk 9]
```diff
@@ -227,6 +231,20 @@ static const char *box_recovery_state_strs[box_recovery_state_MAX] = {
 /** Whether the triggers for "synced" recovery stage have already run. */
 static bool recovery_state_synced_is_reached;
 
+char tarantool_path[PATH_MAX];
+time_t box_start_time;
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/box.cc:235]: Done.

File: src/box/lua/cfg.cc [hunk 10]
```diff
@@ -484,7 +484,7 @@ box_lua_cfg_init(struct lua_State *L)
 		{"cfg_set_replication_anon", lbox_cfg_set_replication_anon},
 		{"cfg_set_net_msg_max", lbox_cfg_set_net_msg_max},
 		{"cfg_set_sql_cache_size", lbox_set_prepared_stmt_cache_size},
-		{"cfg_set_crash", lbox_cfg_set_crash},
+		{"cfg_set_feedback_crash", lbox_cfg_set_feedback_crash},
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/cfg.cc:487]: Done.

File: src/box/box.cc (hunk 1 above)
Comment (nshy) [reply src/box/box.cc:2770]: Done.

File: src/lib/core/crash.h (hunk 2 above)
Comment (nshy) [reply src/lib/core/crash.h:103]: We don't use nanosecond precision actually. Moved this and related change to a distinct patch.

File: src/main.cc (hunk 3 above)
Comment (nshy) [reply src/main.cc:737]: Done. `"failed to get start time, ignore"`

File: src/lib/core/crash.h (hunk 4 above)
Comment (nshy) [reply src/lib/core/crash.h:12]: Fixed.

File: src/lib/core/crash.h (hunk 5 above)
Comment (nshy) [reply src/lib/core/crash.h:22]: Fixed.
```
+#include <stdint.h>
+#include <time.h>
```

File: src/lib/core/crash.h (hunk 6 above)
Comment (nshy) [reply src/lib/core/crash.h:129]: Done.

File: src/main.cc (hunk 7 above)
Comment (nshy) [reply src/main.cc:739]: Done.

File: src/box/box.cc (hunk 8 above)
Comment (nshy) [reply src/box/box.cc:2932]: Issue with reporting crash info on flightrec SIGBUS is fixed.

As to error message - leave "as is" as discussed f2f.

File: src/lib/core/crash.h [hunk 11]
```diff
@@ -5,34 +5,149 @@
...
 
 /**
  * Initialize crash signal handlers.
  */
 void
 crash_signal_init(void);
 
+/**
+ * Report crash information to the stderr (usually a current console).
+ */
```
Comment (locker) [inline src/lib/core/crash.h:147]: Nit: Let's also drop this comment from the function definition in the C file to avoid duplication.

File: src/main.h [hunk 12]
```diff
@@ -39,6 +40,16 @@ extern "C" {
...
 
+/**
+ * Path to Tarantool binary of the process.
+ */
+extern char tarantool_path[PATH_MAX];
+
+/**
+ * Application start time on monotonic clocks.
+ */
+extern time_t start_time_safe;
```
Comment (locker) [inline src/main.h:51]: Nit: `_safe` suffix looks confusing. Also, it'd be nice to have the `tarantool_` prefix for consistency with `tarantool_path`. Let's rename to `tarantool_start_time`?

File: src/main.h [hunk 13]
```diff
@@ -39,6 +40,16 @@ extern "C" {
 
 double tarantool_uptime(void);
 
+/**
+ * Path to Tarantool binary of the process.
+ */
+extern char tarantool_path[PATH_MAX];
```
Comment (locker) [inline src/main.h:46]: Nit: AFAIK you don't need to specify the array length in the declaration. Let's replace with `extern char tarantool_path[]` to eliminate dependency on `limits.h`.

File: src/box/flightrec.h [hunk 14]
```diff
@@ -67,6 +67,17 @@ box_set_flightrec(void)
 	return 0;
 }
 
+/**
+ * This function is called in SIGBUS handler to check whether accessed address
+ * belongs to flightrec file.
+ */
+static inline bool
```
Comment (locker) [inline src/box/flightrec.h:74]: Nit: Let's add `stdbool.h` to the header list.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/lib/core/crash.c [hunk 15]
```diff
@@ -115,18 +115,11 @@ static struct crash_info {
...
-static inline uint64_t
-timespec_to_ns(struct timespec *ts)
-{
-	return (uint64_t)ts->tv_sec * 1000000000 + (uint64_t)ts->tv_nsec;
-}
+static time_t timestamp_mono = 0;
 
 static char *
-ns_to_localtime(uint64_t timestamp, char *buf, ssize_t len)
+ns_to_localtime(time_t sec, char *buf, ssize_t len)
```
[the remark this answers is not in the document]
Comment (locker) [reply src/lib/core/crash.c]: I see that in the next patch you drop this helper function. Maybe, better drop it in the first patch then?

File: src/lib/core/crash.c [hunk 16]
```diff
@@ -90,7 +90,7 @@ static struct crash_info {
 	/**
 	 * Timestamp in nanoseconds (realtime).
 	 */
-	uint64_t timestamp_rt;
+	time_t timestamp_rt;
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/crash.c]: Fixed to `/** Timestamp in seconds (realtime). */` as in next patch.

Changed type to `long` as discussed f2f.

File: src/box/box.cc [hunk 17]
```diff
@@ -4900,6 +5115,7 @@ box_init(void)
 	 * supported from box.cfg not being called yet.
 	 */
 	builtin_events_init();
+	crash_callback = box_crash_callback;
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/box.cc:5125]: Discussed f2f and decided not to reset the callback. This keeps existent behavior of reporting crashes to feedback URL even on application shutdown.

File: src/box/box.h [hunk 18]
```diff
@@ -306,12 +306,17 @@ void box_set_replication_skip_conflict(void);
...
 int box_set_txn_timeout(void);
 int box_set_txn_isolation(void);
 int box_set_auth_type(void);
 int box_set_bootstrap_strategy(void);
 
+/**
+ * Applies box.cfg feedback parameters.
+ * On success, returns 0. On error, sets diag and returns -1.
+ */
+int box_set_feedback(void);
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/box.h:318]: Fixed.

File: src/main.h (hunk 12 above)
Comment (nshy) [reply src/main.h:51]: Done.

File: src/main.h (hunk 13 above)
Comment (nshy) [reply src/main.h:46]: Done.

File: src/box/flightrec.h (hunk 14 above)
Comment (nshy) [reply src/box/flightrec.h:74]: Done.

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
@nshy Compilation fails on a few platforms. Please fix.

**Comment by locker** [comment]
Cherry-picked to 2.10.
